nearcore/src/runtime/mod.rs

```diff
@@ -1049,8 +1054,14 @@ impl RuntimeAdapter for NightshadeRuntime {
         is_new_chunk: bool,
         is_first_block_with_chunk_of_version: bool,
         states_to_patch: SandboxStatePatch,
+        use_flat_storage: bool,
     ) -> Result<ApplyTransactionResult, Error> {
-        let trie = self.get_trie_for_shard(shard_id, prev_block_hash, state_root.clone())?;
+        let trie = self.get_trie_for_shard(
+            shard_id,
+            prev_block_hash,
+            state_root.clone(),
+            use_flat_storage,
+        )?;
```

Comment:
I wonder if passing the `trie` in would be a better solution? I thin we can replace both `state_root` and `use_flat_state` with a single tier arg. 

No opinion on whether that woudl be better or worse

Reply:
Ah it would be but `apply_transactions_with_optional_storage_proof` needs to implemented by both KeyValueRuntime and NightshadeRuntime and KeyValueRuntime does not use trie.

Reply:
The idea is interesting, but it requires creating a `trie` before each `apply_transactions` call, which results in more copy-pasted code for trie creation. We could simplify `apply_transactions_with_optional_storage_proof` interface, but it doesn't have an impact because it is called only once.

I also noticed that `KeyValueRuntime::get_trie_for_shard` actually returns some trie, which is a bit confusing, because this trie is not used for applying transactions. We may want to return an error there, but I can't say for sure because I'm not very familiar with this part of code.